lib/features/support/interactive_cli.rb

```diff
@@ -0,0 +1,162 @@
+
+    @in_stream.close
+
+    maybe_thread = @thread.join(15)
+
+    # The thread did not exit!
+    return false if maybe_thread.nil?
+
+    @pid = nil
+    true
+  end
+
+  # @return [Boolean] Whether the shell is currently running
+  def running?
+    !@pid.nil?
+  end
+
+  # Runs the given command if the shell is running
+  #
+  # @param command [String] The command to run
+  #
+  # @return [Boolean] true if the command is executed, false otherwise
+  def run_command(command)
+    return false unless running?
+
+    @in_stream.puts(command)
+
+    true
+  rescue ::Errno::EIO => err
+    $logger.debug(pid) { "EIO error: #{err}" }
+    false
+  end
+
+  def on_exit(&block)
+    @on_exit_blocks << block
+  end
+
+  private
+
+  # Starts a shell on another thread
+  #
+  # @param shell [String] A path to the shell to run
+  def start_threaded_shell(shell)
+    @thread = Thread.new do
+      start_shell(shell)
+    end
+  end
+
+  # Starts a shell
+  #
+  # @param shell [String] A path to the shell to run
+  def start_shell(shell)
+    stderr_reader, stderr_writer = IO.pipe
+
+    PTY.spawn(shell, err: stderr_writer.fileno) do |stdout, stdin, pid|
+      # We don't need to write to stderr so close it ASAP
+      stderr_writer.close
+
+      $logger.debug(pid) { 'PTY spawned!' }
+      @pid = pid
+      @in_stream = stdin
+
+      stdout_thread = Thread.new do
+        stdout.each_char do |char|
+          if char == "\n"
+            $logger.debug("#{pid} STDOUT") { @current_buffer }
+            @stdout_lines << format_line(@current_buffer)
+            @current_buffer.clear
+          else
+            @current_buffer << char
+          end
+        end
+      rescue ::Errno::EIO => err
+        $logger.debug(pid) { "EIO error: #{err}" }
+      end
+
+      stderr_thread = Thread.new do
+        buffer = ''
+
+        stderr_reader.each_char do |char|
+          if char == "\n"
+            $logger.debug("#{pid} STDERR") { buffer }
+            @stderr_lines << format_line(buffer)
+            buffer.clear
+          else
+            buffer << char
+          end
+        end
+      rescue ::Errno::EIO => err
+        $logger.debug(pid) { "EIO error: #{err}" }
+      end
+
+      _, status = Process.wait2(@pid)
+      @pid = nil
+
+      # Stop the thread that's reading from stdout
+      failed = stdout_thread.join(5).nil?
+      raise 'stdout is blocked!' if failed
+
+      # Stop the thread that's reading from stderr
+      failed = stderr_thread.join(5).nil?
+      raise 'stderr is blocked!' if failed
+
+      $logger.debug(pid) { "PTY exit status: #{status.exitstatus}" }
+      @on_exit_blocks.each do |block|
+        block.call(status.exitstatus)
+      end
+    end
+  ensure
+    stderr_reader.close unless stderr_reader.closed?
+    stderr_writer.close unless stderr_writer.closed?
+  end
+
+  # Strips whitespace from shell lines, can be used to sanitize further if required
+  def format_line(line)
+    line.strip
```

Comment:
Some commands are expected to output with whitespace so I'm not 100% sure this is a good idea; changing it might make things more annoying though

e.g. `git status --porcelain` should output like this:

```
 M some/modified/file.rb
?? a/new/file.rb
```

i.e. modified files are ` M` rather than just `M`, but our assertions have to omit the space or they don't match